### PR TITLE
Improve floating point conversions.

### PR DESCRIPTION
**Pull request by sbertin-telular**
Provides an improved range of conversions and handles scientific
notation.

Updates the tests.

Ignores warning about comparing floating point numbers for equality.
This is intentionally used to detect a NaN.

Signed-off-by: Scott Bertin <sbertin@telular.com>

### CODE REVIEW

**Comment by sbertin-telular** [comment]
Don't merge yet. Further testing showed It is trying to print too many digits in some cases.

**Comment by sbertin-telular** [comment]
Should be good now.

**Comment by dnav** [comment]
Hi Scott,

Thing is scientific notation is not supported in LwM2M text format whereas it is in JSON formats. So could it be possible for utils_floatToText() to have a parameter enabling or disabling the scientific notation ?

From the LwM2M 1.1 TS Core, appendix C:

> Float: Represented as an ASCII signed numeric representation.
For example, we use a floating point number with the significand of 6.667, a base of 10 and the exponent of -11. This represents the number 6.667e-11 in scientific notation and will be represented as "0.00000000006667" as an ASCII string.

File: core/utils.c
Comment (dnav) [inline core/utils.c]: Shouldn't this never be true ?

Comment (sbertin-telular) [reply core/utils.c]: DBL_MAX is the largest finite double value. Infinity will be larger. Infinity needs special handling here or later loops will never terminate.

**Comment by sbertin-telular** [comment]
> Thing is scientific notation is not supported in LwM2M text format whereas it is in JSON formats. So could it be possible for utils_floatToText() to have a parameter enabling or disabling the scientific notation ?

I forgot about text format when doing this. My previous attempt (pull request #419) would be more appropriate for text and this is better for JSON. The two approaches are very different and I think having different functions would be more appropriate. What do you think about reopening #419 and moving this approach to new functions in json.c? We could close this pull request and I'll make a new one for that.

**Comment by sbertin-telular** [comment]
I made it work with a parameter.

**Comment by sbernard31** [comment]
As requested, https://github.com/eclipse/wakaama/issues/482#issuecomment-713653683.

We should begin to integrate current open PR in master.
Regarding @sbertin-telular , this one seems to be the first one to integrate.

@sbertin-telular, could you please rebase this without merge commit (eventually if you think this makes sense you can squash all commit in one).

Then I will integrate it in master.

**Comment by sbertin-telular** [comment]
Rebased and squashed.

**Comment by sbernard31** [comment]
@rettichschnidi,  do you approve this ?

**Review by rettichschnidi — COMMENTED** [review]
First try :)

My findings are not substantial, and most of them clearly have their roots from the way existing code looks. This and given that the current master can not be build without failing tests, I will not insist on any of those findings getting taken care of now. I am also happy contributing some overall improvements (also addressing my findings here) once the PRs up to and including your LWM2M 1.1 branch got merged.

File: tests/CMakeLists.txt
Comment (sbertin-telular) [reply tests/CMakeLists.txt:18]: Not compiling for C90 might be a bug. See #232.

File: tests/convert_numbers_test.c
Comment (sbertin-telular) [reply tests/convert_numbers_test.c:295]: The problem with CU_ASSERT_FATAL is that it prevents later tests from running if it fails. I prefer to know everything that is broken rather than just the first thing. It can help with deciding how to fix it.

File: tests/CMakeLists.txt
Comment (rettichschnidi) [reply tests/CMakeLists.txt:18]: Created #494 to start a discussion about this topic.

File: tests/convert_numbers_test.c
Comment (rettichschnidi) [reply tests/convert_numbers_test.c:295]: Having just one such assertion per test function would help with that. Of course, this would imply another kind of boiler plate... Since the existing code base does not do it, I am OK with leaving it as it is and doing some rework later on, on the whole code base.

File: tests/CMakeLists.txt
Comment (sbernard31) [reply tests/CMakeLists.txt:18]: Eventually and If you both agree we can let's this point aside for this PR. 
I guess maybe #494 could be long discussion and I would not to block the PRs integration for this.
I feel that integrating all the pending work(PRs) is still the priority.

Comment (sbertin-telular) [reply tests/CMakeLists.txt:18]: Thinking about this more I remove my objection. Things like uint8_t are already not part of C90 and are used everywhere.

Comment (sbertin-telular) [reply tests/CMakeLists.txt:18]: Disabling the warning for senml_json.c is still necessary as there is an explicit check for equality to 0.

**Comment by sbertin-telular** [comment]
I forgot to sign off on the pull requests. I'll do a rebase & squash again and make sure that is signed off.

File: core/json.c
Comment (rettichschnidi) [inline core/json.c:839]: Using `isdigit()`/`#include <ctype.h>` makes sense here too.

File: core/json.c
Comment (rettichschnidi) [inline core/json.c:838]: Using `isdigit()`/`#include <ctype.h>` makes sense here too.

**Review by rettichschnidi — CHANGES_REQUESTED** [review]

**Comment by rettichschnidi** [comment]
Found two other places to use `isdigit()` - after this, I am pro merging.

**Review by rettichschnidi — CHANGES_REQUESTED** [review]
> Ignores warning about comparing floating point numbers for equality.
> This is intentionally used to detect a NaN.

This part is not true anymore.

File: core/utils.c
Comment (rettichschnidi) [inline core/utils.c:155]: Using `isdigit()`/`#include <ctype.h>` makes sense here too.

**Review by rettichschnidi — CHANGES_REQUESTED** [review]

File: core/utils.c
Comment (sbertin-telular) [reply core/utils.c:155]: I'll change these as they are new with this pull request, but just grepping for "'0'" shows many more places that isdigit() could be used.

**Comment by sbertin-telular** [comment]
Rebased, squashed, and fixed commit message.

**Comment by sbernard31** [comment]
Integrated in master (commit e6c96710a25c6900f5c5aa1e8f94cd6ee5300267)

@sbertin-telular, you maybe forgot to fetch master before to rebase.

**Pull request closed without merging**
